channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1239 @@
+	return error;
+}
+
+/**
+ * Function description
+ * Estimate RFX_AVC420_BITMAP_STREAM structure size in stream 
+ *
+ * @return estimated size
+ */
+static INLINE UINT32 rdpgfx_estimate_h264_avc420(RDPGFX_AVC420_BITMAP_STREAM* havc420)
+{
+	return sizeof(UINT32) /* numRegionRects */
+		+ (sizeof(RECTANGLE_16) + 2) /* regionRects + quantQualityVals */
+		* havc420->meta.numRegionRects
+		+ havc420->length;
+}
+/**
+ * Function description
+ * Send RDPGFX_CMDID_WIRETOSURFACE_1 or RDPGFX_CMDID_WIRETOSURFACE_2 
+ * message according to codecId
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_send_surface_command(RdpgfxServerContext* context, RDPGFX_SURFACE_COMMAND* cmd)
+{
+	wStream* s;
+	RDPGFX_AVC420_BITMAP_STREAM *havc420 = NULL;
+	RDPGFX_AVC444_BITMAP_STREAM *havc444 = NULL;
+	UINT32 h264Size = 0;
+	UINT32 bitmapDataStart = 0;
+	UINT32 bitmapDataLength = 0;
+
+	/* Create new stream according to codec. */
+	if (cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE ||
+		cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE_V2)
+	{
+		s = rdpgfx_server_packet_new(
+		    RDPGFX_CMDID_WIRETOSURFACE_2, 
+		    sizeof(RDPGFX_WIRE_TO_SURFACE_PDU_2) + cmd->length);
+	}
+	else if (cmd->codecId == RDPGFX_CODECID_AVC420)
+	{
+		havc420 = (RDPGFX_AVC420_BITMAP_STREAM *)cmd->extra;
+		h264Size = rdpgfx_estimate_h264_avc420(havc420);
+
+		s = rdpgfx_server_packet_new(
+		    RDPGFX_CMDID_WIRETOSURFACE_1, 
+		    sizeof(RDPGFX_WIRE_TO_SURFACE_PDU_1) + h264Size);
+	}
+	else if (cmd->codecId == RDPGFX_CODECID_AVC444)
+	{
+		havc444 = (RDPGFX_AVC444_BITMAP_STREAM *)cmd->extra;
+		h264Size = sizeof(UINT32); /* cbAvc420EncodedBitstream1 */
+
+		/* avc420EncodedBitstream1 */
+		havc420 = &(havc444->bitstream[0]);
+		h264Size += rdpgfx_estimate_h264_avc420(havc420);
+
+		/* avc420EncodedBitstream2 */
+		if (havc444->LC == 0)
+		{
+			havc420 = &(havc444->bitstream[1]);
+			h264Size += rdpgfx_estimate_h264_avc420(havc420);
+		}
+
+		s = rdpgfx_server_packet_new(
+		    RDPGFX_CMDID_WIRETOSURFACE_1, 
+		    sizeof(RDPGFX_WIRE_TO_SURFACE_PDU_1) + h264Size);
+	}
+	else
+	{
+		s = rdpgfx_server_packet_new(
+		    RDPGFX_CMDID_WIRETOSURFACE_1, 
+		    sizeof(RDPGFX_WIRE_TO_SURFACE_PDU_1) + sizeof(cmd->length));
+	}
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rdpgfx_server_packet_new failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	/* Reformat and send */
+	if (cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE ||
+	    cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE_V2)
+	{
+		Stream_Write_UINT16(s, cmd->surfaceId); /* surfaceId (2 bytes) */
+		Stream_Write_UINT16(s, cmd->codecId); /* codecId (2 bytes) */
+		Stream_Write_UINT32(s, cmd->contextId); /* codecContextId (4 bytes) */
+		Stream_Write_UINT8(s, cmd->format); /* pixelFormat (1 byte) */
+
+		Stream_Write_UINT32(s, cmd->length); /* bitmapDataLength (4 bytes) */
+		Stream_Write(s, cmd->data, cmd->length);
+
+		return rdpgfx_server_packet_send(context, s);
+	}
+	else
+	{
```

Comment:
This else block can be moved one level down (as the if above returns) and the "else" removed